processing/src/main/java/org/apache/carbondata/processing/loading/sort/unsafe/UnsafeSortDataRows.java

```diff
@@ -141,13 +141,18 @@ public void initialize() throws MemoryException {
     semaphore = new Semaphore(parameters.getNumberOfCores());
   }
 
-  private UnsafeCarbonRowPage createUnsafeRowPage() throws MemoryException {
+  private UnsafeCarbonRowPage createUnsafeRowPage()
+      throws MemoryException, CarbonSortKeyAndGroupByException {
     MemoryBlock baseBlock =
         UnsafeMemoryManager.allocateMemoryWithRetry(this.taskId, inMemoryChunkSize);
     boolean isMemoryAvailable =
         UnsafeSortMemoryManager.INSTANCE.isMemoryAvailable(baseBlock.size());
     if (isMemoryAvailable) {
       UnsafeSortMemoryManager.INSTANCE.allocateDummyMemory(baseBlock.size());
+    } else {
+      LOGGER.info("trigger in-memory merge and spill for table " + parameters.getTableName());
```

Comment:
FYI: The main entry point is here

Reply:
I think this approach is good for case that the sort memory size is smaller than the input data size. For example, sort memory is 10GB and input data size is 25GB, then it is good. But in case sort memory is 10GB and input size is 15GB, then the new approach will spill more data to disk, so current approach way is better.
So why not let user to configure how much data to spill.

Reply:
Oh, I think the configuration will be difficult to understand. If the user configured 10GB, the spilled data files may be less since the file is compressed.

Besides, we have to verify this configuration to make sure that it is less than sort memory.

Most importantly, the sort memory is shared by all data loading. A configuration of this may not tell how it will affect the real behavior.

I prefer to keep it simple now and the result is quite obvious: In the worst scenario, we only have to spill extra #sizeOfSortMemory data to disk comparing with the previous implementation. And in general scenario, we can spill bigger files to disk instead of more smaller files.